pipeline/src/pipeline.rs

```diff
@@ -787,14 +791,35 @@ impl<T: FieldElement> Pipeline<T> {
         Ok(self.artifact.analyzed_asm.as_ref().unwrap())
     }
 
+    pub fn compute_optimized_asm(&mut self) -> Result<AnalysisASMFile, Vec<String>> {
+        if let Some(ref optimized_asm) = self.artifact.optimized_asm {
+            return Ok(optimized_asm.clone());
+        }
+
+        self.compute_analyzed_asm()?;
+        let analyzed_asm = self.artifact.analyzed_asm.take().unwrap();
+
+        self.log("Optimizing asm...");
+        let optimized = powdr_asmopt::optimize(analyzed_asm);
+        self.maybe_write_pil(&optimized, "_asm_opt")?;
+
+        self.artifact.optimized_asm = Some(optimized);
+
+        Ok(self.artifact.optimized_asm.as_ref().unwrap().clone())
+    }
+
+    pub fn optimized_asm(&self) -> Result<AnalysisASMFile, Vec<String>> {
+        Ok(self.artifact.optimized_asm.as_ref().unwrap().clone())
+    }
+
     pub fn compute_constrained_machine_collection(
         &mut self,
     ) -> Result<&AnalysisASMFile, Vec<String>> {
         if self.artifact.constrained_machine_collection.is_none() {
             self.artifact.constrained_machine_collection = Some({
-                self.compute_analyzed_asm()?;
-                let analyzed_asm = self.artifact.analyzed_asm.take().unwrap();
-                powdr_asm_to_pil::compile::<T>(analyzed_asm)
+                self.compute_optimized_asm()?;
```

Comment:
this already returns the `optimized_asm`

Reply:
Yes, but I understand that this way you avoid having to clone the file. It's similar to how it's done for the PIL optimization pass.